### PR TITLE
docs(api): document more public expression APIs

### LINKED ISSUE #5263: docs: missing docs for some public expression APIs
### What happened?

Some functions have no description at all, so they are not visible in the documentation.
Screenshot 2023-01-19 at 11 59 45
This is a waste of their potential, because only a handful of developers will know about them and will want to review the code and then see how it has been implemented.

```shell
$ git ls-files | grep '^ibis/expr/types' | xargs -t ruff --isolated --select=D102 --show-source
ibis/expr/types/core.py:66:9: D102 Missing docstring in public method
   |
66 |     def equals(self, other):
   |         ^^^^^^ D102
   |

ibis/expr/types/core.py:78:9: D102 Missing docstring in public method
   |
78 |     def has_name(self):
   |         ^^^^^^^^ D102
   |

ibis/expr/types/core.py:81:9: D102 Missing docstring in public method
   |
81 |     def get_name(self):
   |         ^^^^^^^^ D102
   |

ibis/expr/types/core.py:175:9: D102 Missing docstring in public method
    |
175 |     def op(self) -> ops.Node:
    |         ^^ D102
    |

ibis/expr/types/generic.py:62:9: D102 Missing docstring in public method
   |
62 |     def type(self):
   |         ^^^^ D102
   |

ibis/expr/types/generic.py:485:9: D102 Missing docstring in public method
    |
485 |     def to_projection(self) -> ir.Table:
    |         ^^^^^^^^^^^^^ D102
    |

ibis/expr/types/generic.py:512:9: D102 Missing docstring in public method
    |
512 |     def as_table(self) -> ir.Table:
    |         ^^^^^^^^ D102
    |

ibis/expr/types/generic.py:781:9: D102 Missing docstring in public method
    |
781 |     def first(self) -> Column:
    |         ^^^^^ D102
    |

ibis/expr/types/generic.py:784:9: D102 Missing docstring in public method
    |
784 |     def last(self) -> Column:
    |         ^^^^ D102
    |

ibis/expr/types/generic.py:787:9: D102 Missing docstring in public method
    |
787 |     def rank(self) -> Column:
    |         ^^^^ D102
    |

ibis/expr/types/generic.py:790:9: D102 Missing docstring in public method
    |
790 |     def dense_rank(self) -> Column:
    |         ^^^^^^^^^^ D102
    |

ibis/expr/types/generic.py:793:9: D102 Missing docstring in public method
    |
793 |     def percent_rank(self) -> Column:
    |         ^^^^^^^^^^^^ D102
    |

ibis/expr/types/generic.py:796:9: D102 Missing docstring in public method
    |
796 |     def cume_dist(self) -> Column:
    |         ^^^^^^^^^ D102
    |

ibis/expr/types/generic.py:801:9: D102 Missing docstring in public method
    |
801 |     def cummin(self) -> Column:
    |         ^^^^^^ D102
    |

ibis/expr/types/generic.py:804:9: D102 Missing docstring in public method
    |
804 |     def cummax(self) -> Column:
    |         ^^^^^^ D102
    |

ibis/expr/types/generic.py:807:9: D102 Missing docstring in public method
    |
807 |     def lag(
    |         ^^^ D102
    |

ibis/expr/types/generic.py:814:9: D102 Missing docstring in public method
    |
814 |     def lead(
    |         ^^^^ D102
    |

ibis/expr/types/generic.py:821:9: D102 Missing docstring in public method
    |
821 |     def ntile(self, buckets: int | ir.IntegerValue) -> ir.IntegerColumn:
    |         ^^^^^ D102
    |

ibis/expr/types/groupby.py:87:9: D102 Missing docstring in public method
   |
87 |     def aggregate(self, metrics=None, **kwds):
   |         ^^^^^^^^^ D102
   |

ibis/expr/types/groupby.py:292:9: D102 Missing docstring in public method
    |
292 |     def summary(self, exact_nunique=False):
    |         ^^^^^^^ D102
    |

ibis/expr/types/groupby.py:301:9: D102 Missing docstring in public method
    |
301 |     def summary(self, exact_nunique=False):
    |         ^^^^^^^ D102
    |

ibis/expr/types/logical.py:80:9: D102 Missing docstring in public method
   |
80 |     def any(self, where: BooleanValue | None = None) -> BooleanValue:
   |         ^^^ D102
   |

ibis/expr/types/logical.py:85:9: D102 Missing docstring in public method
   |
85 |     def notany(self, where: BooleanValue | None = None) -> BooleanValue:
   |         ^^^^^^ D102
   |

ibis/expr/types/logical.py:90:9: D102 Missing docstring in public method
   |
90 |     def all(self, where: BooleanValue | None = None) -> BooleanScalar:
   |         ^^^ D102
   |

ibis/expr/types/logical.py:93:9: D102 Missing docstring in public method
   |
93 |     def notall(self, where: BooleanValue | None = None) -> BooleanScalar:
   |         ^^^^^^ D102
   |

ibis/expr/types/logical.py:96:9: D102 Missing docstring in public method
   |
96 |     def cumany(self) -> BooleanColumn:
   |         ^^^^^^ D102
   |

ibis/expr/types/logical.py:99:9: D102 Missing docstring in public method
   |
99 |     def cumall(self) -> BooleanColumn:
   |         ^^^^^^ D102
   |

ibis/expr/types/numeric.py:468:9: D102 Missing docstring in public method
    |
468 |     def cummean(self) -> NumericColumn:
    |         ^^^^^^^ D102
    |

ibis/expr/types/numeric.py:489:9: D102 Missing docstring in public method
    |
489 |     def cumsum(self) -> NumericColumn:
    |         ^^^^^^ D102
    |

ibis/expr/types/relations.py:96:9: D102 Missing docstring in public method
   |
96 |     def as_table(self):
   |         ^^^^^^^^ D102
   |

ibis/expr/types/relations.py:248:9: D102 Missing docstring in public method
    |
248 |     def columns(self):
    |         ^^^^^^^ D102
    |

Found 31 error(s).
```

### What version of ibis are you using?

dev

### What backend(s) are you using, if any?

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
This PR documents more public expression APIs. I did not document deprecated methods (`to_projection`) or methods that are public but not typically used by analytics end users (`op`). Closes #5263.

### CODE REVIEW

**Comment by krzysztof-kwitt** [comment]
Can we add pre-commit check to avoid regression? It shouldn't be hard to add using ruff linter.

File: ibis/expr/types/logical.py [hunk 1]
```diff
@@ -78,23 +78,35 @@ class BooleanScalar(NumericScalar, BooleanValue):
 @public
 class BooleanColumn(NumericColumn, BooleanValue):
     def any(self, where: BooleanValue | None = None) -> BooleanValue:
+        """Return whether elements are `True`.
```
Comment (krzysztof-kwitt) [inline ibis/expr/types/logical.py:81]:
```suggestion
        """Return whether at least one element are `True`.
```

**Comment by kszucs** [comment]
Could you please add scope to the commit?

File: ibis/expr/types/core.py [hunk 2]
```diff
@@ -76,9 +95,11 @@ def __bool__(self) -> bool:
     __nonzero__ = __bool__
 
     def has_name(self):
+        """Check whether this expression has an explicit name."""
         return isinstance(self._arg, ops.Named)
 
     def get_name(self):
+        """Return this name of this expression."""
```
Comment (kszucs) [inline ibis/expr/types/core.py:102]:
```suggestion
        """Return the name of this expression."""
```

File: ibis/expr/types/groupby.py [hunk 3]
```diff
@@ -290,14 +291,11 @@ def __init__(self, arr, parent):
     group_concat = _group_agg_dispatch('group_concat')
 
     def summary(self, exact_nunique=False):
+        """Summarize a column of non-numeric values."""
```
Comment (kszucs) [inline ibis/expr/types/groupby.py:294]: Can't there be numeric values?

Comment (cpcloud) [reply ibis/expr/types/groupby.py:294]: Yep, fixing.

File: ibis/expr/types/core.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/types/core.py]: Fixed.

File: ibis/expr/types/groupby.py (hunk 3 above)
Comment (cpcloud) [reply ibis/expr/types/groupby.py]: Fixed.

**Comment by cpcloud** [comment]
> Could you please add scope to the commit?

Done.

File: ibis/expr/types/logical.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/types/logical.py]: Thank you! Fixed.

**Comment by cpcloud** [comment]
> Can we add pre-commit check to avoid regression? It shouldn't be hard to add using ruff linter.

Some of these functions I'd rather not document, like `op`. I also don't want `D102` to run over the whole codebase. Is there a way to invoke `ruff` to satisfy both of those?

**Comment by krzysztof-kwitt** [comment]
@cpcloud What do you think about it?
```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 3d40a44f5..8ce21f5a3 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -26,7 +26,11 @@ repos:
     hooks:
       - id: ruff
         # exclude a file (if configured to do so), even if it's passed in explicitly
-        args: ["--force-exclude", "--fix"]
+        args: ["--force-exclude"]
+      - id: ruff
+        # exclude a file (if configured to do so), even if it's passed in explicitly
+        args: ["--force-exclude", '--isolated', '--select=D102', '--show-source']
+        files: ^ibis/expr/types
   - repo: https://github.com/adrienverge/yamllint
     rev: v1.29.0
     hooks:
diff --git a/ibis/expr/types/core.py b/ibis/expr/types/core.py
index 8249a2dfd..9fd7482e0 100644
--- a/ibis/expr/types/core.py
+++ b/ibis/expr/types/core.py
@@ -193,7 +193,7 @@ class Expr(Immutable):
         else:
             return f(self, *args, **kwargs)
 
-    def op(self) -> ops.Node:
+    def op(self) -> ops.Node:  # noqa: D102
         return self._arg
 
     def _find_backends(self) -> tuple[list[BaseBackend], bool]:
diff --git a/ibis/expr/types/generic.py b/ibis/expr/types/generic.py
index 487fa9af5..b47c22128 100644
--- a/ibis/expr/types/generic.py
+++ b/ibis/expr/types/generic.py
@@ -483,7 +483,7 @@ class Value(Expr):
         return ops.SortKey(self, ascending=False).to_expr()
 
     @util.deprecated(as_of="4.1", removed_in="5.0", instead="use `.as_table()`")
-    def to_projection(self) -> ir.Table:
+    def to_projection(self) -> ir.Table:  # noqa: D102
         return self.as_table()
 
     def as_table(self) -> ir.Table:
diff --git a/pyproject.toml b/pyproject.toml
index e5682c416..c7aabb9f1 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -394,6 +394,10 @@ target-version = "py38"
 
 [tool.ruff.per-file-ignores]
 "*test*.py" = ["D"] # ignore all docstring lints in tests
+"ibis/expr/types/**" = [
+  "D", # ignore all docstring lints in tests
+  "RUF100" # Unused `noqa` directive
+]
 
 [tool.black]
 line_length = 88

```

**Comment by krzysztof-kwitt** [comment]
You can also create rseperate uff configuration in `ibis/expr/types` directory.
https://github.com/charliermarsh/ruff#pyprojecttoml-discovery

**Comment by krzysztof-kwitt** [comment]
or you can also add RUF100 to noqa
```
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 3d40a44f5..bf0f6a503 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -27,6 +27,10 @@ repos:
       - id: ruff
         # exclude a file (if configured to do so), even if it's passed in explicitly
         args: ["--force-exclude", "--fix"]
+      - id: ruff
+        # exclude a file (if configured to do so), even if it's passed in explicitly
+        args: ["--force-exclude", '--isolated', '--select=D102', '--show-source']
+        files: ^ibis/expr/types
   - repo: https://github.com/adrienverge/yamllint
     rev: v1.29.0
     hooks:
diff --git a/ibis/expr/types/core.py b/ibis/expr/types/core.py
index 8249a2dfd..8b03bd37c 100644
--- a/ibis/expr/types/core.py
+++ b/ibis/expr/types/core.py
@@ -193,7 +193,7 @@ class Expr(Immutable):
         else:
             return f(self, *args, **kwargs)
 
-    def op(self) -> ops.Node:
+    def op(self) -> ops.Node:  # noqa: D102, RUF100
         return self._arg
 
     def _find_backends(self) -> tuple[list[BaseBackend], bool]:
diff --git a/ibis/expr/types/generic.py b/ibis/expr/types/generic.py
index 487fa9af5..2afdcbe77 100644
--- a/ibis/expr/types/generic.py
+++ b/ibis/expr/types/generic.py
@@ -483,7 +483,7 @@ class Value(Expr):
         return ops.SortKey(self, ascending=False).to_expr()
 
     @util.deprecated(as_of="4.1", removed_in="5.0", instead="use `.as_table()`")
-    def to_projection(self) -> ir.Table:
+    def to_projection(self) -> ir.Table:  # noqa: D102, RUF100
         return self.as_table()
 
     def as_table(self) -> ir.Table:
diff --git a/pyproject.toml b/pyproject.toml
index e5682c416..c7aabb9f1 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -394,6 +394,10 @@ target-version = "py38"
 
 [tool.ruff.per-file-ignores]
 "*test*.py" = ["D"] # ignore all docstring lints in tests
+"ibis/expr/types/**" = [
+  "D", # ignore all docstring lints in tests
+  "RUF100" # Unused `noqa` directive
+]
 
 [tool.black]
 line_length = 88

```

**Review by krzysztof-kwitt — APPROVED** [review]

**Pull request merged**